clang/test/Sema/cheri/init-list-narrowing.c

```diff
@@ -0,0 +1,105 @@
+  {(void * __capability)42},
+  // hybrid-c-warning@-1 {{incompatible pointer to integer conversion initializing 'int' with an expression of type 'void * __capability __attribute__((cheri_no_provenance))'}}
+  // purecap-c-warning@-2 {{incompatible pointer to integer conversion initializing 'int' with an expression of type 'void * __attribute__((cheri_no_provenance))'}}
+  // hybrid-cxx-error@-3 {{cannot initialize a member subobject of type 'int' with an rvalue of type 'void * __capability __attribute__((cheri_no_provenance))'}}
+  // purecap-cxx-error@-4 {{cannot initialize a member subobject of type 'int' with an rvalue of type 'void * __attribute__((cheri_no_provenance))'}}
+  {(void * __capability)foo},
+  // hybrid-c-warning@-1 {{incompatible pointer to integer conversion initializing 'int' with an expression of type 'void * __capability'}}
+  // purecap-c-warning@-2 {{incompatible pointer to integer conversion initializing 'int' with an expression of type 'void *'}}
+  // hybrid-cxx-error@-3 {{cannot initialize a member subobject of type 'int' with an rvalue of type 'void * __capability'}}
+  // purecap-cxx-error@-4 {{cannot initialize a member subobject of type 'int' with an rvalue of type 'void *'}}
+  {(__intcap)0},
+  {(__intcap)42},
+#ifdef __cplusplus
+  {__null},
+  {(__intcap)__null},
+#endif
+};
+
+struct {
+  void *p;
+} ptrs[] = {
+  {(void *)0},
+  {(void *)42},
+  {(void *)foo},
+  {(void * __capability)0},
+  // hybrid-cxx-error@-1 {{converting capability type 'void * __capability' to non-capability type 'void *' without an explicit cast; if this is intended use __cheri_fromcap}}
+  {(void * __capability)42},
+  // hybrid-c-error@-1 {{converting capability type 'void * __capability __attribute__((cheri_no_provenance))' to non-capability type 'void *' without an explicit cast; if this is intended use __cheri_fromcap}}
+  // hybrid-cxx-error@-2 {{converting capability type 'void * __capability' to non-capability type 'void *' without an explicit cast; if this is intended use __cheri_fromcap}}
+  {(void * __capability)foo},
+  // hybrid-error@-1 {{converting capability type 'void * __capability' to non-capability type 'void *' without an explicit cast; if this is intended use __cheri_fromcap}}
+  {(__intcap)0},
+  // expected-c-warning@-1 {{expression which evaluates to zero treated as a null pointer constant of type 'void *'}}
+  // expected-cxx-error@-2 {{cannot initialize a member subobject of type 'void *' with an rvalue of type '__intcap __attribute__((cheri_no_provenance))'}}
+  {(__intcap)42},
+  // expected-c-warning@-1 {{incompatible integer to pointer conversion initializing 'void *' with an expression of type '__intcap __attribute__((cheri_no_provenance))'}}
+  // expected-cxx-error@-2 {{cannot initialize a member subobject of type 'void *' with an rvalue of type '__intcap __attribute__((cheri_no_provenance))'}}
```

Comment:
Could you add a few more cases here that should result in a C++ narrowing warning by using an out-of-range constant:
```
extern char foo[];
#define __capability
typedef long __intcap;
constexpr __intcap l = (long)__INT_MAX__ + 1;
constexpr int i = __INT_MIN__;

struct {
  int x;
} ints[] = {
  {(__intcap)(__LONG_MAX__)},
  {(__intcap)&foo[0]},
  {l},
  {(__intcap)i},
};
```
https://godbolt.org/z/7dexja1Gq

Reply:
Added a bunch of `(long)__INT_MAX__ + 1` tests; don't particularly want to be exhaustively testing the boundary cases, just that we can get that diagnostic out